### ISSUE TITLE
Proposal: New option for alignment of virtual diagnostic text

### ISSUE THREAD

**Issue by berggeist** [yegappan/lsp#260]
I would like to propose a new option for the alignment of the virtual diagnostic text.

Currently the default behavior, if the existing option `showDiagWithVirtualText` is set to `v:true`, is that the diagnostic virtual text messages are displayed above the related lines, as can be seen in the following example:
![lsp-virtual-text-align-above](https://user-images.githubusercontent.com/24396556/232216748-807f2182-01be-44bf-82b7-e956b5494662.png)
In this case all diagnostic messages are truncated and each diagnostic message is displayed in its own virtual text line.

---

Now the proposal is, to introduce an additional option `diagVirtualTextAlign` that can take one of the following values:
- `above` , the default value
- `after`
- `below`

---

`above` is the default value if the new option `diagVirtualTextAlign` is not customized. In this case, the currently implemented behavior is preserved. An example is already given above.

---

`after` will display the virtual diagnostic text messages in the same line but after the related line of text. Here is an example:
![lsp-virtual-text-align-after](https://user-images.githubusercontent.com/24396556/232218143-840aeddb-f387-41b8-bcd3-fb45324a104a.png)
Now multiple diagnostic messages are joined within one virtual text line and prefixed by the according symbol for the severity. As the joined messages might become rather long, in this case, the virtual text is not truncated but wrapped.

---

`below` is more or less the opposite to `above`. I.e. in this case, the virtual text for the diagnostic messages are aligned below the related line of text:
![lsp-virtual-text-align-below](https://user-images.githubusercontent.com/24396556/232219249-9b105d4d-7d74-4fa6-ab4a-dea01e665d46.png)

---

I would like to hear your opinion, whether this proposal would be helpful and whether you agree that I create a PR for this new feature.

**Comment by yegappan** [comment]
I think this option will be useful for some users who prefer the multiple diagnostics to be shown after the line.  Can you go ahead and create a PR for this?

**Comment by andlrc** [comment]
> `below` is more or less the opposite to `above`. I.e. in this case, the virtual text for the diagnostic messages are aligned below the related line of text:
> ![lsp-virtual-text-align-below](https://user-images.githubusercontent.com/24396556/232219249-9b105d4d-7d74-4fa6-ab4a-dea01e665d46.png)

This is just nitpick, but shouldn't the two diagnostics be flipped around, so that "Cannot find name 'ddd' is rendered above "Variable 'ccc' implicitly ...."

**Comment by berggeist** [comment]
@yegappan : Thanks for your reply. I'm going to prepare a PR.

@andlrc: No it's not a nitpick. The same came to my mind, too. There is a function `SortDiags` that sorts the diagnostic messages according to their columns positions within the same line. I've just now modified this function by adding an additional argument for reversing the sort order, in the case `below` has been chosen for the alignment.
Now it looks like that:
![lsp-vim-virtual-text-align-below-rev-columns](https://user-images.githubusercontent.com/24396556/232230978-33c28d97-539f-4c19-bcd5-d658a913c3ab.png)

Does it now look better? I'm wondering, how much this re-sorting of diagnostics might slow-down this mechanism for large lists of diagnostics.
Maybe it would be the best I provide two PR's:
- the basic modifications for my first proposal
- a second PR for the correction of the sort order in the case `below` is used for alignment.

**Comment by Shane-XB-Qian** [comment]
> I think this option will be useful for some users who prefer the multiple diagnostics to be shown after the line. 

perhaps it better be at `above`, compl menu normally would stay at `below`, 
so IMO align perhaps no need to change, but adding a limitation of the diag messages amount, if it's over 2, perhaps just display the next one as `there xxx msg left in this line`, otherwise too many msg there.

**Comment by Shane-XB-Qian** [comment]
or the limitation (max number of diag msg be displayed) amount was that option.
// if you align the virtual diag text at `below`, i guess it would conflict with compl popup etc.

**Comment by berggeist** [comment]
@Shane-XB-Qian : The current behavior, .i.e. `above` is still the default and will be completely preserved as is.
The new configuration option `diagVirtualTextAlign` will just provide additional alignment options.

Personally, I prefer the option `after`, as for me the current situation with `above` was a little bit irritating, because depending on the availability of diagnostics in the currently edited line suddenly a new (virtual text) line pops up and shifts the cursor position up  and down.

But your suggestion for limiting the number of diagnostic messages  sounds also interesting. I'll try to provide a PR for this option, too. Would `diagVirtualTextMaxMessages` (of type number) be an acceptable name for this additional option?

**Comment by andlrc** [comment]
> @andlrc: No it's not a nitpick. The same came to my mind, too. There is a function `SortDiags` that sorts the diagnostic messages according to their columns positions within the same line. I've just now modified this function by adding an additional argument for reversing the sort order, in the case `below` has been chosen for the alignment.
> Now it looks like that:

Keep in mind that `sortDiags` are expensive, and therefore only runs once when diags arrive. Also keep in mind that the sorted diagnostics are also used for commands like `:LspDiagNext`. I'm not sure what the best approach would be? Possible push the prop config to an intermediate list and add them in reverse at the end? Or possible whenever the lnum changes?

**Comment by Shane-XB-Qian** [comment]
> Personally, I prefer the option `after`, as for me the current situation with `above` was a little bit irritating, because depending on the availability of diagnostics in the currently edited line suddenly a new (virtual text) line pops up and shifts the cursor position up and down.

ok, i see, just `below` perhaps much worse, :smile: 



> But your suggestion for limiting the number of diagnostic messages sounds also interesting. I'll try to provide a PR for this option, too. Would `diagVirtualTextMaxMessages` (of type number) be an acceptable name for this additional option?

i think better be, otherwise there are 3/4.. (and more) diag messages of virtual text displayed for a line, looks awkward.
// just gave a summary if over that max number.

// as for the option name, i preferred `Num` as its suffix, but this is on @yegappan

**Comment by berggeist** [comment]
@andlrc Yes, that is exactly what I've been worrying about and this left me to the conclusion that it's not worth to re-sort the messages for alignment below.

In addition with with the reply of @Shane-XB-Qian  who mentions that with `below` the diagnostic messages would conflict with completions I`m wondering whether `below` will be used very often.

Nevertheless I'm going to provide multiple different PR's and ask you all to review them thoroughly.

**Comment by andlrc** [comment]
> Possible push the prop config to an intermediate list and add them in reverse at the end?

@berggeist ^ What I mean by this is something like what is implemented in this diff:

```diff
diff --git a/autoload/lsp/diag.vim b/autoload/lsp/diag.vim
index 1d28a40..e07a471 100644
--- a/autoload/lsp/diag.vim
+++ b/autoload/lsp/diag.vim
@@ -142,6 +142,7 @@ def DiagsRefresh(bnr: number)
   endif
 
   var signs: list<dict<any>> = []
+  var virtualTextProps: list<list<any>>
   var diags: list<dict<any>> = diagsMap[bnr].sortedDiagnostics
   for diag in diags
     # TODO: prioritize most important severity if there are multiple diagnostics
@@ -166,11 +167,11 @@ def DiagsRefresh(bnr: number)
         if padding > 0
           padding = strdisplaywidth(getline(diag.range.start.line + 1)[ : diag.range.start.character - 1])
         endif
-        prop_add(lnum, 0, {bufnr: bnr,
+        virtualTextProps->add([lnum, 0, {bufnr: bnr,
                            type: 'LspDiagVirtualText',
                            text: $'┌─ {diag.message}',
                            text_align: 'above',
-                           text_padding_left: padding})
+                           text_padding_left: padding}])
       endif
     catch /E966\|E964/ # Invalid lnum | Invalid col
       # Diagnostics arrive asynchronous and the document changed while they wore
@@ -178,6 +179,11 @@ def DiagsRefresh(bnr: number)
     endtry
   endfor
 
+  # Add ->reverse() when position == 'below'
+  for virtualTextProp in virtualTextProps
+    function('prop_add', virtualTextProp)()
+  endfor
+
   signs->sign_placelist()
 enddef
 
```

**Comment by berggeist** [comment]
Hi @andlrc thank you very much for your information and the provided patch!

That makes it clear, that with your solution any side-effects concerning the sort-order of `diagsMap[bnr].sortedDiagnistics` will be eliminated. On the other side I'm not so happy that the complete list of virtual text properties would be reversed, because generally I'd expect the diagnostics might contain many lines, and only few of them include multiple diagnostics.

I would rather favor a line based approach where (hopefully) only those lines need to be reversed, which contain multiple diagnostics.
I peeked into the available source code and found that there is already a dict `serverDiagnosticsByLnum` and a function `GetDiagsByLine` maybe both might be helpful for such a line-based approach.

**Comment by andlrc** [comment]
I assume this can be closed?

**Comment by berggeist** [comment]
Solved by the merge of #273
